haze/src/androidMain/kotlin/dev/chrisbanes/haze/RenderScriptBlurEffect.kt

```diff
@@ -0,0 +1,196 @@
+
+  private val density: Density
+    get() = node.requireDensity()
+
+  override fun DrawScope.drawEffect() {
+    val offset = node.layerOffset
+    var scaleFactor = node.calculateInputScaleFactor()
+    var blurRadiusPx = scaleFactor * with(density) { node.resolveBlurRadius().toPx() }
+
+    if (blurRadiusPx > MAX_BLUR_RADIUS) {
+      // RenderScript has a max blur radius (25px), so to create an equivalent visual effect
+      // we need to increase the scale factor
+      scaleFactor *= (MAX_BLUR_RADIUS / blurRadiusPx)
+      blurRadiusPx = MAX_BLUR_RADIUS
+    }
+
+    HazeLogger.d(TAG) { "drawEffect. blurRadius=${blurRadiusPx}px. scaleFactor=$scaleFactor" }
+
+    val scaledLayer = createScaledContentLayer(
+      node = node,
+      scaleFactor = scaleFactor,
+      layerSize = node.layerSize,
+      layerOffset = offset,
+    )
+    if (scaledLayer != null) {
+      if (drawGraphicsLayer == null) {
+        // If there is no graphics layer yet, then this is the first draw. We'll generate
+        // this blocking, so that the user doesn't see an un-blurred first frame
+        runBlocking {
+          updateSurface(content = scaledLayer, blurRadius = blurRadiusPx)
+          // Release the graphics layer
+          graphicsContext.releaseGraphicsLayer(scaledLayer)
+        }
+      } else {
+        node.coroutineScope.launch(Dispatchers.Main.immediate) {
+          updateSurface(content = scaledLayer, blurRadius = blurRadiusPx)
+          // Release the graphics layer
+          graphicsContext.releaseGraphicsLayer(scaledLayer)
+        }
+      }
+    }
+
+    drawGraphicsLayer?.let { layer ->
+      drawScaledContentLayer(offset = -offset, scaleFactor = scaleFactor) {
+        // TODO: apply mask or progressive to layer?
+        // TODO: draw noise?
+
+        layer.alpha = node.alpha
+        drawLayer(layer)
+
+        if (node.alpha < 1f) {
+          PaintPool.usePaint { paint ->
+            paint.alpha = node.alpha
+            drawContext.canvas.withSaveLayer(size.toRect(), paint) {
+              for (tint in node.resolveTints()) {
+                drawScrim(mask = node.mask, progressive = node.progressive, tint = tint)
+              }
+            }
+          }
+        } else {
+          for (tint in node.resolveTints()) {
+            drawScrim(mask = node.mask, progressive = node.progressive, tint = tint)
+          }
+        }
+      }
+    }
+  }
+
+  private suspend fun updateSurface(content: GraphicsLayer, blurRadius: Float) {
+    val rs = getRenderScriptContext(
+      context = node.currentValueOf(LocalContext),
+      size = content.size,
+    )
+    // Draw the layer (this is async)
+    rs.inputSurface.drawGraphicsLayer(layer = content, density = density, drawScope = drawScope)
+    // Wait for the layer to be written to the Surface
+    rs.awaitSurfaceWritten()
+
+    if (!node.isAttached) return
+
+    // Now apply the blur
+    rs.applyBlur(blurRadius)
```

Comment:
By the way, @chrisbanes, 
I noticed this change when you flattened the callback for setOnBufferAvailableListener. You moved the blurring process to the **main thread**. 
Previously, when I was calling the `applyBlur` function in the `onDataReceived` callback, the callback itself was running on a separate RenderScript-created thread called something like `RenderScriptMessageThread`. The duration of the blurring process can vary from device to device. On my test device with a Snapdragon 888, it took around 5 to 10 milliseconds according to my measurements. That’s almost half a frame at 60 Hz, or a full frame on a 120 Hz screen.

Reply:
Good catch, fixed. 

I'm avoiding using RenderScript's message thread, as that is a thread for messaging, not computing/effects. Ended up just using `Dispatcher.Default`